src/mpid/ch4/src/ch4_progress.c

```diff
@@ -14,6 +16,37 @@ static int MPIDI_Progress_made(MPID_Progress_state * state)
     return (state->progress_count != state->progress_start);
 }
 
+static int get_progress_vci(MPID_Progress_state * state)
+{
+    int vci;
+
+    if (state->vci_count == 0) {
+        /* global progress, state->vci[] not needed */
+        vci = state->vci_idx++;
+        if (state->vci_idx == MPIDI_global.n_vcis) {
+            state->vci_idx = 0;
+        }
+    } else if ((++state->poll_count) == MPIDI_CH4_GLOBAL_POLL_FREQ) {
+        /* time to poll global progress */
+        state->poll_count = 0;
+        vci = MPL_atomic_fetch_add_int(&MPIDI_global.next_global_vci, 1);
+        if (vci == MPIDI_global.n_vcis - 1) {
+            MPL_atomic_store_int(&MPIDI_global.next_global_vci, 0);
+        }
+    } else if (MPIDI_global.n_vcis == 1) {
+        /* single vci */
+        vci = state->vci[0];
+    } else {
+        /* several vcis */
+        vci = state->vci[state->vci_idx++];
+        if (state->vci_idx == state->vci_count) {
+            state->vci_idx = 0;
+        }
+    }
```

Comment:
I think that encoding the global vs. per-vci progress information in the `state` variables is unclear and also leads to more branches. It seems like separating out the type of progress information with, for example, `MPIDI_PROGRESS_TYPE__PER_VCI` and `MPIDI_PROGRESS_TYPE__GLOBAL` is clearer; it also incurs fewer branches.

Reply:
Passing in the information with explicit parameter or inside the `state` structure is equivalent, right? I prefer to do it with the state structure since we are already using the state and the opaqueness of state structure give us more flexibility in adding or removing parameters.